### DISCUSSION TITLE
Feature Request: list_sort with lambda `list_sort(list_of_structs, x -> x.age)`

### DISCUSSION THREAD

**Discussion by NickCrews** [duckdb/duckdb#10417] [Ideas]
I have a list of structs. I want them sorted by a certain field within the struct. Consider:

```SQL
SELECT list_sort([{"name": 'b', "age": 1}, {"name": 'c', "age": 3}, {"name": 'a', "age": 2}]);
```

I want them sorted by the age. This could look like `list_sort(list_of_structs, x -> x.age)`. The lambda is optional. This is the same semantics as python's sort function.

As a workaround at present, I noticed that the above query results in `[{"name": "a", "age": 2}, {"name": "b", "age": 1}, {"name": "c", "age": 3}] `, which makes me think that duckdb currently sorts on the first first field in the struct. I can take advantage of this behavior to make a wrapper (this could be done directly in duckdb too, but I'm using ibis for clarity) and get this behavior with no needed changes in duckdb:

```python
from ibis.expr.types import ArrayValue
from typing import Callable

def sort(arr: ArrayValue, *, key: Callable | None = None):
    if key is None:
        return arr.sort()
    wrapped = arr.map(lambda el: ibis.struct({"key": key(el), "value": el}))
    sorted = wrapped.sort()
    unwrapped = sorted.map(lambda wrapped: wrapped.value)
    return unwrapped
```

but it would be great if this was builtin to duckdb directly

1. Is it part of duckdb's official behavior that it sorts list<struct> by the first field? If so, this should be documented.
2. Can we make this be built in to list_sort()?

**Comment by NickCrews** [comment]
@cpcloud FYI this is might have an effect on Ibis. If I sort an array of structs in ibis, the order might depend on the backend. Ibis should probably have a policy here, or explicitly say that sort order of structs depends on the backend. Or it could have the wrapper I posted above as a compat layer for different backends.

**Reply by cpcloud** [reply]
Most backends don't have great support for sorting structs, so at most we're likely to document it.
